Initial doc, scripts and sample app (#2)
Azure-Samples/azure-edge-extensions-iothub-fileupload-privatelink: README.md

# Setting up Azure IoT Hub File Upload to Azure Storage through Private Endpoint

How to enable Azure IoT Hub's file upload functionality through an Azure Storage only allowing private connections.

The file upload functionality provided by Azure IoT Hub serves as a secure bridge to the Azure Storage account, eliminating the need for IoT client devices to rely on external custom services for secure connections to the Storage account. For a comprehensive understanding of this feature, refer to the official  [Azure IoT Hub file upload overview](https://learn.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload).

This repo offers guidance on securing Azure Storage account through a private endpoint while still allowing IoT Hub to interact with it securely, as well as allowing client IoT devices to reach the account through a gateway (Application Gateway and optionally Azure Firewall). With this approach the Storage account no longer allows public Internet access, which is a common requirement in enterprise deployments.

## Features

> This project provides different options for configuring Azure IoT Hub File upload to a Storage private endpoint:

Reviewer: ```suggestion
This project provides different options for configuring Azure IoT Hub file upload to an Azure Storage private endpoint:
```

> * [Documentation](./docs/doc-iothub-storage-private.md) with guidance an details on how you can accomplish this configuration in a Hub-Spoke network topology, leveraging Application Gateway, and Azure Firewall for traffic inspection

Reviewer: ```suggestion
* [Documentation](./docs/doc-iothub-storage-private.md) with guidance and details on how you can accomplish this configuration in a [hub-spoke network topology](https://learn.microsoft.com/azure/architecture/networking/architecture/hub-spoke), using Azure Application Gateway, and Azure Firewall for traffic inspection.
```

> * Quickstart: Azure CLI based step-by-step scripts with a simplified setup (single Virtual Network & Resource Group) and sample client app

Reviewer: ```suggestion
* Quickstart: Azure CLI based step-by-step scripts with a simplified setup (single virtual network & resource group) and sample client application.
```

## Getting Started

### Prerequisites

> * Azure subscription

Reviewer: Please include what level of permissions are needed as well on that subscription.

* Azure CLI
* Visual Studio Code
* OpenSSL CLI
* Bash terminal, on Windows you can use WSL
* .NET SDK 8 (if you wish to run the client and server samples for file upload)

> * DNS provider if you wish to test HTTPS with your custom domain, and SSL Certificate for your custom domain

Reviewer: Please link all of these their respective install instructions.

> [!TIP]
> Without a custom DNS and an SSL certificate available, you can still test the entire setup but the IoT client device portion of this quickstart will not be completed.

### Quickstart

This quickstart provides instructions on how to set up Azure IoT Hub's file upload functionality through an Azure Storage account that only allows private connections, through step-by-step scripts. The scripts deploy and configure a simplified set of Azure resources to showcase Azure Storage with private endpoint, Application Gateway and Azure IoT Hub routing to Azure Storage. The Storage account disables any public Internet access and only accessible within the Virtual Network. In this quickstart there is no Azure Firewall for traffic inspection.
Finally, two sample .NET apps interact with the resources deployed to showcase the end to end data flows.

> [!WARNING]

> > This sample currently generates self-signed certificate and key files on disk. This should be improved to leverage Azure Key Vault for securely storing and retrieving the password, certificate and key. Application Gateway supports retrieving an SSL certificate from a Key Vault account.

Reviewer: ```suggestion
> This sample currently generates self-signed certificate and key files on disk. This must be improved to use Azure Key Vault for securely storing and retrieving the password, certificate and key before using this in production. Application Gateway supports retrieving an SSL certificate from Azure Key Vault.
```

> #### Setup

Reviewer: Please add a visual at the top of this file to help a learner understand what they are going to be deplying here.  While you have some of that information in text, a visual will be very helpful.

* Open a Bash terminal
* `git clone` this repo

> * cd [repository name]

Reviewer: Use code fence blocks for commands.  in this case, it would be something like...

~~~suggestion
1. Clone this repo and go to the root directory.

   ```bash
   git clone https://github.com/Azure-Samples/azure-edge-extensions-iothub-fileupload-privatelink
   cd azure-edge-extensions-iothub-fileupload-privatelink
   ```
~~~

* Open the project in Visual Studio Code `code .` and open a Bash terminal.

> * Log into your Azure account `az login [--tenant xxxx-xxx]` and set the default subscription `az account set -s <subscription>`

Reviewer: Steps should always be numbered, not bullets.  Bullets are for unordered lists, steps are ordered actions.

* Prepare required environment variables to run the deployment scripts. We create a file in the folder `./temp` which is excluded from Git.

  * Run the below to create the file `./temp/envvars.sh`, for `LOCATION` you can choose any Azure region supporting IoT Hub and Application Gateway.

    ```bash
    if [ ! -d "./temp" ]; then
        mkdir ./temp
    fi
    >./temp/envvars.sh cat <<EOF
    # change the below set to match your environment based on Readme
    export TENANT_ID="xxx-xxxx-xxx-xxxx-xxxx"

>     export CERT_PASSWORD="xxx"

Reviewer: Please don't demonstrate, even in development, even with a fake cert, a cert password in a file.  Have the script prompt the user as part of the deployment steps.  This is an SFI (teach security) requirement.

Author: This is no longer part of the repo, it's from an early version

    export LOCATION="westeurope"
    export PREFIX="xxx"
    export RESOURCE_GROUP="rg-xxx-xxx"
    EOF

    code ./temp/envvars.sh
    ```

  * The newly created `.sh` file should now open in Visual Studio Code.
  * Edit the values to your preference, ensure `TENANT_ID` corresponds to your Azure tenant.
  * For `PREFIX` use a short 5 character string that can be unique for the resource name composition.
  * Load the variables by running:

  ```bash
  source ./temp/envvars.sh
  ```

#### Azure Resources with Virtual Network, private Storage and Application Gateway

Deploy the Azure components for setting up Virtual Network, self-signed SSL certificate, Storage, Private Link, custom private DNS and Application Gateway configured to talk to the Storage account. This allows you to validate the flow before configuring Azure IoT Hub and client communication.

* From the root directory of this repo, run the first part of the deployment. The script will use the environment variables and build composed resource names by appending the `PREFIX` variable as Azure resource names.

```bash
./deploy/quickstart-agw-storage.sh
```

> * It will take a few minutes to deploy all resources. Keep the terminal open and take note of some of the generated DNS entries.

Reviewer: Again, steps need to be numbered

* This scripts also uploads a `sample.txt` file to a Blob storage container and generates a SAS URI token for testing.
* From a bash terminal, try out a `CURL` command to the DNS of the Public IP address attached to the Application Gateway.

```bash
curl --insecure "<APP_GATEWAY_SASURI copied output>"
```

* Run the same command with the Blob Public URI printed out by the script `BLOB_SASURI` value. Verify this does not succeed since this Storage account is blocking direct Internet traffic.

> * Configure your custom DNS and SSL certificate for end to end SSL encryption.

Reviewer: The instructions below use a publicly accessible Key Vault to store the Pfx. You need to call out the security concerns with an internet-facing Key Vault that doesn't use private endpoints as a production warning.   (SFI teaching opportunity)
